src/context/WalletProvider/Ledger/components/Chains.tsx

```diff
@@ -0,0 +1,161 @@
+
+      setLoadingChains(prevLoading => ({ ...prevLoading, [chainId]: true }))
+
+      try {
+        const { chainNamespace } = fromChainId(chainId)
+        const chainIds = chainId === ethChainId ? getSupportedEvmChainIds() : [chainId]
+        const accountMetadataByAccountId = await deriveAccountIdsAndMetadataForChainNamespace[
+          chainNamespace
+        ]({
+          accountNumber: 0,
+          chainIds,
+          wallet: walletState.wallet,
+        })
+
+        const accountIds = Object.keys(accountMetadataByAccountId)
+        const { getAccount } = portfolioApi.endpoints
+        const opts = { forceRefetch: true }
+        const accountPromises = accountIds.map(accountId =>
+          dispatch(getAccount.initiate({ accountId }, opts)),
+        )
+        const accountResults = await Promise.allSettled(accountPromises)
+
+        const balanceByChainId = accountResults.reduce<Record<ChainId, BN>>((acc, res, idx) => {
+          if (res.status === 'rejected') return acc
+          const { data: account } = res.value
+          if (!account) return acc
+
+          const accountId = accountIds[idx]
+          const { chainId } = fromAccountId(accountId)
+          const accountBalance = Object.values(account.accountBalances.byId).reduce(
+            (acc, byAssetId) => {
+              Object.values(byAssetId).forEach(balance => (acc = acc.plus(bnOrZero(balance))))
+              return acc
+            },
+            bnOrZero(0),
+          )
+
+          acc[chainId] = bnOrZero(acc[chainId]).plus(accountBalance)
+
+          const accountMetadata = accountMetadataByAccountId[accountId]
+          const payload = { [accountId]: accountMetadata }
+          dispatch(portfolio.actions.upsertAccountMetadata(payload))
+          dispatch(portfolio.actions.upsertPortfolio(account))
+          return acc
+        }, {})
+
+        Object.entries(balanceByChainId).forEach(([chainId, balance]) => {
+          if (balance.eq(0)) pull(availableChainIds, chainId)
+        })
+      } catch (e) {
+        console.error(e)
+      } finally {
+        setLoadingChains(prevLoading => ({ ...prevLoading, [chainId]: false }))
+      }
+    },
+    [availableChainIds, dispatch, walletState.adapters, walletState.wallet],
+  )
+
+  return (
+    <>
+      <ModalBody textAlign='left' pb={8}>
+        <Box mb={4}>
+          <Text
+            fontSize='lg'
+            fontWeight='bold'
+            translation={'walletProvider.ledger.chains.header'}
+            mb={2}
+          />
+          <Text color='text.subtle' translation={'walletProvider.ledger.chains.body'} />
+        </Box>
+        <Box mb={4}>
+          <Text
+            fontSize='md'
+            fontWeight='bold'
+            translation={'walletProvider.ledger.chains.availableChains'}
+            mb={2}
+          />
+          <Box>
+            {availableAssets
+              .sort(a => (walletChainIds.includes(a.chainId) ? 1 : -1))
```

Comment:
would it be worth memoizing the sorting of availableAssets?

Reply:
8521dd1d5529399074082153de1460e46bcd3b42